sopht_mpi/sopht_mpi_simulator/flow/flow_simulators_mpi_2d.py

```diff
@@ -0,0 +1,299 @@
+import numpy as np
+from sopht_mpi.numeric.eulerian_grid_ops import (
+    gen_advection_timestep_euler_forward_conservative_eno3_pyst_mpi_kernel_2d,
+    gen_diffusion_timestep_euler_forward_pyst_mpi_kernel_2d,
+    gen_penalise_field_boundary_pyst_mpi_kernel_2d,
+    gen_outplane_field_curl_pyst_mpi_kernel_2d,
+    gen_update_vorticity_from_velocity_forcing_pyst_mpi_kernel_2d,
+    UnboundedPoissonSolverMPI2D,
+)
+from sopht_mpi.utils import MPIConstruct2D, MPIGhostCommunicator2D
+from sopht.numeric.eulerian_grid_ops import (
+    gen_add_fixed_val_pyst_kernel_2d,
+    gen_set_fixed_val_pyst_kernel_2d,
+)
+from sopht.utils.precision import get_test_tol
+from mpi4py import MPI
+
+
+class UnboundedFlowSimulator2D:
+    """Class for MPI-supported 2D unbounded flow simulator"""
+
+    def __init__(
+        self,
+        grid_size,
+        x_range,
+        kinematic_viscosity,
+        CFL=0.1,
+        flow_type="passive_scalar",
+        with_free_stream_flow=False,
+        real_t=np.float32,
+        rank_distribution=None,
+        **kwargs,
+    ):
+        """Class initialiser
+
+        :param grid_size: Grid size of simulator
+        :param x_range: Range of X coordinate of the grid
+        :param kinematic_viscosity: kinematic viscosity of the fluid
+        :param CFL: Courant Freidrich Lewy number (advection timestep)
+        :param flow_type: Nature of the simulator, can be "passive_scalar" (default value),
+        "navier_stokes" or "navier_stokes_with_forcing"
+        :param with_free_stream_flow: has free stream flow or not
+        :param real_t: precision of the solver
+        :param rank_distribution: distribution configuration of the grid
+
+        Notes
+        -----
+        Currently only supports Euler forward timesteps :(
+        """
+        self.grid_dim = 2
+        self.grid_size = grid_size
+        self.grid_size_y, self.grid_size_x = self.grid_size
+        self.x_range = x_range
+        self.real_t = real_t
+        self.flow_type = flow_type
+        self.with_free_stream_flow = with_free_stream_flow
+        self.kinematic_viscosity = kinematic_viscosity
+        self.CFL = CFL
+        supported_flow_types = [
+            "passive_scalar",
+            "navier_stokes",
+            "navier_stokes_with_forcing",
+        ]
+        if self.flow_type not in supported_flow_types:
+            raise ValueError("Invalid flow type given")
+        if self.flow_type == "passive_scalar" and self.with_free_stream_flow:
+            raise ValueError(
+                "Free stream flow not defined for passive advection diffusion!"
+            )
+
+        # MPI-related variables
+        self.rank_distribution = rank_distribution
+        # How do we know max(ghost_size) among all employed kernels apriori
+        # before we even compile the kernels? Hardcoding it for now.
+        self.ghost_size = 2
+
```

Comment:
let's not do it this way. We have a check against this anyways, in the check valid ghost size inside kernel compilation. Instead, let's catch the exception `ValueError` inside the kernel compilation and raise an error that is more legible on the simulator level. As for the value of ghost size, let's set it to default value =2 in the keyword args.

Reply:
Sounds good actually. I have implemented `ghost_size` as an args in [4f3c094](https://github.com/fankiat/sopht-mpi/pull/53/commits/4f3c094c7c92a5ee030deadb31e860e7dda90515) and also make the error messages more explicit to print out the name of the kernel being compiled in [1a663ca](https://github.com/fankiat/sopht-mpi/pull/53/commits/1a663ca1c32cee7b215359b670de558e2ff69f5b).

Reply:
9c64d9d